### PR TITLE
Move knapsack algorithms into separate filing

### PR DESCRIPTION
**Pull request by basilwong**
Summary:
Original Context Doc: https://docs.google.com/document/d/1Gv5ZqN3UY7kTuCUd0JLU3L_onwVIr2vd3AiZyim1Muc/edit?disco=AAABZX_tqdk

### Changes

This diff restructures the Partitioners.py file in the _functorch package.

* Moves the three knapsack problem algorithems (greedy, ilp, dp) into a separate file

Test Plan:
### Unit Testing

```
$ buck test mode/opt //caffe2/test/functorch:test_ac
File changed: fbsource//xplat/caffe2/test/functorch/TARGETS
File changed: fbsource//xplat/caffe2/test/functorch
File changed: fbsource//xplat/caffe2/test
7 additional file change events
Soft Error: source_directory_includes_subpackage: Directory `v2.17.1-1` of package `fbsource//third-party/nccl` may not cover any subpackages, but includes subpackage `v2.17.1-1/src/tests`.
Soft Error: source_directory_includes_subpackage: Directory `v2.18.3-1` of package `fbsource//third-party/nccl` may not cover any subpackages, but includes subpackage `v2.18.3-1/src/tests`.
Soft Error: source_directory_includes_subpackage: Directory `v2.19.3-1` of package `fbsource//third-party/nccl` may not cover any subpackages, but includes subpackage `v2.19.3-1/src/tests`.
Buck UI: https://www.internalfb.com/buck2/a2f91f8a-5326-435e-9075-5af0de930b8b
Test UI: https://www.internalfb.com/intern/testinfra/testrun/7036874660108924
Network: Up: 28MiB  Down: 3.5GiB  (reSessionID-19af3d71-7528-448c-9126-5615d27b3bd7)
Jobs completed: 423656. Time elapsed: 3:57.2s.
Cache hits: 99%. Commands: 146147 (cached: 145758, remote: 317, local: 72)
Tests finished: Pass 8. Fail 0. Fatal 0. Skip 0. Build failure 0

```

### Tested on Local Training Run

```
CUDA_VISIBLE_DEVICES=5,6 AOT_PARTITIONER_DEBUG=1 PARTITIONER_MEMORY_BUDGET_PARETO=0 buck2 run mode/opt //aps_models/ads/icvr:icvr_launcher -- mode=local_fb_fm_v4 launcher.num_workers=2 2>&1 | tee log_2024-11-2320:41:50.757256__bento_trigger.txt
```

Output Summary Paste: P1685697066

Differential Revision: D65800097

### CODE REVIEW

File: torch/_functorch/_activation_checkpointing/knapsack.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    memory: List[float], runtimes: List[float], max_memory: float
+) -> Tuple[float, List[int], List[int]]:
+    import numpy as np
+
+    try:
+        from scipy.optimize import Bounds, LinearConstraint, milp
+    except ImportError:
+        raise RuntimeError(
+            "To use the ILP for memory budget checkpointing you need to install scipy"
+        ) from None
```
Comment (Skylion007) [inline torch/_functorch/_activation_checkpointing/knapsack.py:40]: Ugh, why doesn't this just raise from the ImportError. Really unclear why erasing all the useful import error data from the trace is useful.

**Review by jansel — CHANGES_REQUESTED** [review]
Tests failing

**Comment by basilwong** [comment]
> Tests failing

Hmm interesting. This diff shouldn't be introducing any logic changes. Let me look into why the tests are failing.

File: torch/_functorch/_activation_checkpointing/knapsack.py (hunk 1 above)
Comment (basilwong) [reply torch/_functorch/_activation_checkpointing/knapsack.py:42]: This PR is just moving logic from the partitioners file to a separate file. 

Maybe @Chillee can answer your question. Logic was written as part of https://github.com/pytorch/pytorch/pull/126320?fbclid=IwZXh0bgNhZW0CMTEAAR19wrq3QY7vibee0B_mabKSFaxr54DE6EXml4sxvjEMDTcbPE9UsoL2aDE_aem_n0yhwybp_d4eiz4LAiv7JQ

**Comment by basilwong** [comment]
Closing/unlinking this PR. 

```
13Error: Credentials could not be loaded, please check your action inputs: Could not load credentials from any providers
```

This error in the signals is caused because this PR was exported from diff as a fork. 

The remediation is the export another PR not a fork after unlinking from old pr.

**Pull request closed without merging**